### PR TITLE
Simplify GH Actions config with matrix syntax

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

Our config is very repetitive. Let's try to DRY it so that it's easier to maintain.

In the commit history, I first demonstrate how you can split a build with a matrix of variables. The manual split took ~15 minutes to run all specs. Then I show how to use Knapsack to do the splitting for us and that took ~12 minutes.

~~*Update:* The last build took only 9m 41s in total.~~
*Update:* The last build took 11 minutes in total.

Knapsack will need rebalancing from time to time to make it more efficient but that's easier done than manually re-balancing specs. The manual split also required some extra code to make sure that we weren't forgetting any specs. Knapsack is doing all that for us.

#### What should we test?

I checked that we run the same amount of specs as before. You can double-check if you like. Otherwise, good to go if the specs pass.

#### Release notes

Changelog Category: Technical changes

Simplified configurations of automated tests.

### CODE REVIEW

**Comment by jibees** [comment]
I assume the version with Knapsack can be a good idea: we don't manager ourself the optimization and it's done by a product designed fo this purpose.
I'm wondering if price of this solution isn't a blocking point?

**Comment by mkllnk** [comment]
> I'm wondering if price of this solution isn't a blocking point?

Knapsack is free. There is KnapsackPro which we would need to pay for but I don't think that we need it.

**Comment by andrewpbrett** [comment]
@mkllnk I resolved a merge conflict for this. It does look like there's a spec that's failing consistently though, and I didn't see an obvious explanation.

**Comment by mkllnk** [comment]
Hm, it's passing when run on its own. The error suggests that it may have been caused by the recent rubocop PR that Luis submitted, introducing frozen string literals everywhere. Master is passing though. It may be order dependent... 

Running the same set of specs fails on my machine, too:
```
CI_NODE_INDEX="8" CI_NODE_TOTAL=10 bundle exec rake 'knapsack:rspec[--format progress --tag ~performance]'
```

In case this is an order dependent failure, I'm trying to find the conflicting set of specs:
```
r --bisect --tag ~performance --pattern "spec/features/consumer/shopping/variant_overrides_spec.rb,spec/controllers/admin/subscriptions_controller_spec.rb,spec/features/consumer/shopping/cart_spec.rb,spec/models/spree/return_authorization_spec.rb,spec/models/spree/ability_spec.rb,spec/controllers/line_items_controller_spec.rb,spec/controllers/spree/admin/orders_controller_spec.rb,spec/lib/open_food_network/packing_report_spec.rb,spec/requests/checkout/stripe_sca_spec.rb,spec/services/order_fees_handler_spec.rb,spec/services/invoice_renderer_spec.rb,spec/features/admin/enterprises/images_spec.rb,spec/requests/api/orders_spec.rb,spec/controllers/checkout_controller_spec.rb,spec/features/admin/order_cycles/complex_editing_spec.rb,spec/helpers/enterprises_helper_spec.rb,spec/services/order_workflow_spec.rb,spec/features/consumer/caching/darkswarm_caching_spec.rb,spec/features/consumer/shopping/products_spec.rb,spec/controllers/admin/schedules_controller_spec.rb,spec/serializers/api/admin/order_serializer_spec.rb,spec/views/spree/orders/edit.html.haml_spec.rb,spec/features/admin/configuration/content_spec.rb,spec/models/spree/shipping_method_spec.rb,spec/features/admin/order_cycles/complex_editing_exchange_same_enterprise_spec.rb,spec/models/enterprise_fee_spec.rb,spec/lib/tasks/data/truncate_data_rake_spec.rb,spec/models/spree/order/checkout_spec.rb,spec/helpers/checkout_helper_spec.rb,spec/models/calculator/weight_spec.rb,spec/services/order_payment_finder_spec.rb,spec/models/spree/gateway/stripe_sca_spec.rb,spec/serializers/api/variant_serializer_spec.rb,spec/services/process_payment_intent_spec.rb,spec/models/spree/user_spec.rb,spec/views/spree/admin/payment_methods/index.html.haml_spec.rb,spec/serializers/api/admin/subscription_line_item_serializer_spec.rb,spec/queries/outstanding_balance_spec.rb,spec/services/checkout/post_checkout_actions_spec.rb,spec/models/concerns/balance_spec.rb,spec/helpers/admin/orders_helper_spec.rb,spec/models/spree/order/tax_spec.rb,spec/controllers/user_confirmations_controller_spec.rb,spec/lib/open_food_network/property_merge_spec.rb,spec/mailers/user_mailer_spec.rb,spec/controllers/api/v0/terms_and_conditions_controller_spec.rb,spec/lib/tasks/enterprises_rake_spec.rb,spec/controllers/api/v0/states_controller_spec.rb,spec/models/calculator/flexi_rate_spec.rb,spec/helpers/shop_helper_spec.rb,spec/services/variant_units/option_value_namer_spec.rb,spec/controllers/api/v0/enterprise_fees_controller_spec.rb,spec/models/product_import/settings_spec.rb,spec/lib/spree/core/mail_settings_spec.rb,spec/services/default_stock_location_spec.rb,spec/models/spree/order/address_spec.rb,spec/services/content_sanitizer_spec.rb,spec/lib/spree/localized_number_spec.rb,spec/lib/open_food_network/order_grouper_spec.rb,spec/models/tag_rule/filter_products_spec.rb,spec/controllers/spree/admin/countries_controller_spec.rb,spec/models/subscription_line_item_spec.rb,spec/serializers/api/credit_card_serializer_spec.rb,spec/services/permitted_attributes/order_cycle_spec.rb,spec/models/spree/preferences/configuration_spec.rb,spec/models/calculator/flat_rate_spec.rb,spec/helpers/serializer_helper_spec.rb,spec/migrations/migrate_enterprise_fee_tax_amounts_spec.rb"
```

Unfortunately, this seems to run for a long time. The initial knapsack run took only 5 minutes but this is longer for over 8 minutes now. I will see...

I tried running knapsack with the `--bisect` command but rspec complained that the spec order wasn't deterministic that way. I'll continue this tomorrow.

**Comment by mkllnk** [comment]
I got `rspec --bisect` running in a reasonable time by deleting all specs except the ones that were run in the failing knapsack run. Then I could run:
```
bundle exec rspec --bisect=verbose --tag ~performance --seed 0
...

The minimal reproduction command is:
  rspec ./spec/features/consumer/shopping/cart_spec.rb[1:1:3:1] ./spec/migrations/migrate_enterprise_fee_tax_amounts_spec.rb[1:1:1:1] --tag ~performance --seed 0
```

And that seems to be correct. It reproduces the fail locally. I can now look into the cart spec to find out why it's breaking the tax amounts spec.

**Comment by mkllnk** [comment]
Still learning about `rspec --bisect`. So apparently it works well with knapsack but we need to give it a seed to have a defined order. Maybe it was just luck but this worked as well:
```
CI_NODE_INDEX="8" CI_NODE_TOTAL=10 bundle exec rake 'knapsack:rspec[--format progress --tag ~performance --bisect --seed 0]'
```

**Comment by mkllnk** [comment]
This combination of specs is failing in master as well. It's failing on this branch because the spec order is random. It could happen on any branch now.

**Comment by mkllnk** [comment]
The offending line is:
```rb
require_relative '../../db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts'
```

Loading the migration code breaks the application somehow.

**Comment by mkllnk** [comment]
More specifically, this line seems to be the problem:
https://github.com/openfoodfoundation/openfoodnetwork/blob/e37a065f10c107a543ed05df2ff1a8e723a8f826/db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb#L19

By referencing the model name `Spree::Variant`, we load that model instead of using the model defined in the migration. The migration's model is `MigrateEnterpriseFeeTaxAmounts::Spree::Variant`. I can confirm that by explicitly referencing the migration's model, the failure disappears. I will try to fix the migration or spec and submit a PR. Otherwise we just have to deactivate the spec. We don't need to run a spec for a migration that will never be run again.

**Comment by mkllnk** [comment]
@Matt-Yorkley I identified a problem with loading `db/migrate/20210406161242_migrate_enterprise_fee_tax_amounts.rb` in specs. Loading that code can fail other specs. I first thought that I could encapsulate the class names but that breaks Rails' polymorphism. You found the same problem in the past. Now I'm wondering if you have any better idea than removing that spec so that we don't load the migration code. What do you think?

Code to reproduce:
```
rspec ./spec/features/consumer/shopping/cart_spec.rb[1:1:3:1] ./spec/migrations/migrate_enterprise_fee_tax_amounts_spec.rb[1:1:1:1]
```

**Review by andrewpbrett — APPROVED** [review]
This looks like a much nicer approach! Removing the migration spec seems fine to me but I'd defer to what @Matt-Yorkley thinks.

**Comment by Matt-Yorkley** [comment]
I encountered this before and left a warning here: https://github.com/openfoodfoundation/openfoodnetwork/commit/c702b398d61e05f28cc2b09303a70a00ef8521bf#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721R275-R278

TLDR; bad things happen if you run migration tests alongside other tests, I think due to a combination of `good_migrations` and redefining `ActiveModel` classes in migrations. Isolating them in their own task in the build is a good idea.

As a side note: we can delete that specific migration spec because it's already served it's purpose. I'll be adding some different migration specs soon though, and having the ability to add and run them without things exploding is fairly useful..?

**Comment by mkllnk** [comment]
> I'll be adding some different migration specs soon though, and having the ability to add and run them without things exploding is fairly useful..?

Yes, that would be very useful. Several thoughts:

* Migrations that don't need to load polymorphic associations can probably be written without loading app code.
* We could exclude all migration specs from rspec runs by default:
  `config.filter_run_excluding migration: true`.
   But I don't see how you can then run them manually. We could have that line depending on an ENV var.
* We could exclude migration specs from CI runs with `--tag ~migration`.
* After the release has been rolled out, we don't need to run the migration's specs any more.
* We could name migration specs with the migration's timestamp and only run the last (few) specs.
* We could tag specs with the db version and run only those specs for the current version:
  `rspec --tag "db$(rake db:version | cut -d ' ' -f 3)"`
* :shrug:

**Comment by Matt-Yorkley** [comment]
Agreed on all points, including the shrug :smile:

**Comment by mkllnk** [comment]
Great. So do you approve of this pull request, @Matt-Yorkley?

**Comment by Matt-Yorkley** [comment]
We would still have a broken build next time we add a migration spec though, right?

**Comment by mkllnk** [comment]
> We would still have a broken build next time we add a migration spec though, right?

No, only if that spec loads app code.

My approach here is:

- Delete the current migration spec because it's not needed any more.
- Implement some of my ideas above (e.g. run separately) when needed for a future migration spec.

That would mean that we can go ahead now and implement new stuff as needed. The alternative is to invest the time now and create a setup to run that migration spec so that we have an example how to do it even though we don't need it at the moment.

I prefer to avoid that complexity now but I am aware that it could mean that the learnings from this get lost over time. What do you think?
